nixos/modules/services/misc/ollama.nix

```diff
@@ -255,14 +260,44 @@ in
           ];
           UMask = "0077";
         };
-      postStart = mkBefore ''
-        set -x
-        export OLLAMA_HOST=${lib.escapeShellArg cfg.host}:${builtins.toString cfg.port}
-        for model in ${lib.escapeShellArgs cfg.loadModels}
-        do
-          ${lib.escapeShellArg (lib.getExe ollamaPackage)} pull "$model"
-        done
-      '';
+    };
+
+    systemd.services.ollama-model-loader = lib.mkIf ((builtins.length cfg.loadModels) > 0) {
+      description = "Download ollama models in the background";
+      wantedBy = [
+        "multi-user.target"
+        "ollama.service"
+      ];
+      after = [ "ollama.service" ];
+      bindsTo = [ "ollama.service" ];
+      environment = config.systemd.services.ollama.environment;
+      serviceConfig = {
+        Type = "exec";
+        DynamicUser = true;
+        Restart = "on-failure";
+        # bounded exponential backoff
+        RestartSec = "1s";
+        RestartMaxDelaySec = "2h";
+        RestartSteps = "10";
+      };
+
+      script =
+        let
+          loadModels = builtins.map (
+            model: "'${lib.getExe ollamaPackage}' pull ${lib.escapeShellArg model} &"
+          ) cfg.loadModels;
+        in
+        ''
+          ${lib.concatStringsSep "\n" loadModels}
```

Comment:
I assumed an unrolled loop would be faster in bash, since my understanding is that bash is exceptionally slow, even worse than python, and thus loops should be avoided when possible. Is that inaccurate? Irrelevant in this case?

Reply:
Bash definitely isn't fast, but I'd not expect any benefit from "unrolled loops", and besides you won't really notice bash when you're downloading gigabyte-sized models

Reply:
May be worth the notice, there's quite a bit of recursion and loops in stdenv: https://nixos.org/manual/nixpkgs/unstable/#ssec-stdenv-dependencies-propagated

Reply:
I changed the script, see if it's more readable now. It still doesn't read the models from argv or an environment variable, though, so it may still not be how you want.